File: ts/src/parse-env.ts
/**
 * A small, forgiving `.env` parser.
 *
 * Supports `KEY=VALUE`, `export KEY=VALUE`, `#` comments, blank lines, and
 * single- or double-quoted values (with `\n`, `\t`, `\r`, `\\`, `\"` escapes
 * inside double quotes). Line numbers are preserved for diagnostics.
 */

/** A parsed assignment from a `.env` file. */
export interface EnvEntry {
  readonly key: string;
  readonly value: string;
  /** 1-based line number the assignment was found on. */
  readonly line: number;
}

/**
 * An error encountered while parsing a `.env` file. Carries the offending
 * 1-based line number so callers can point users at the exact spot.
 */
export class ParseError extends Error {
  readonly line: number;

  constructor(line: number, message: string) {
    super(`line ${line}: ${message}`);
    this.name = "ParseError";
    this.line = line;
  }
}

/** Parse the contents of a `.env` file into a list of entries. */
export function parseEnv(contents: string): EnvEntry[] {
  const entries: EnvEntry[] = [];
  const lines = contents.split("\n");

  for (let idx = 0; idx < lines.length; idx++) {
    // `String.split("\n")` yields a trailing "" for content ending in a
    // newline; Rust's `.lines()` does not. Skipping blank/comment lines below
    // makes the two behave identically without special-casing.
    const rawLine = stripCarriageReturn(lines[idx] ?? "");
    const line = idx + 1;
    const trimmedStart = trimStart(rawLine);
    if (trimmedStart.length === 0 || trimmedStart.startsWith("#")) {
      continue;
    }

    const withoutExport = trimmedStart.startsWith("export ")
      ? trimmedStart.slice("export ".length)
      : trimmedStart;

    const eq = withoutExport.indexOf("=");
    if (eq < 0) {
      throw new ParseError(line, `missing '=' in assignment: ${quote(rawLine)}`);
    }

    const key = withoutExport.slice(0, eq).trim();
    if (key.length === 0 || !isValidKey(key)) {
      throw new ParseError(line, `invalid variable name: ${quote(key)}`);
    }

    const value = parseValue(withoutExport.slice(eq + 1).trim(), line);
    entries.push({ key, value, line });
  }

  return entries;
}

/**
 * Parse a `.env` file's contents into a map, keeping the last assignment when
 * a key is repeated. Throws {@link ParseError} (with a line number) if the
 * file is malformed.
 */
export function envFromDotenv(contents: string): Record<string, string> {
  const map: Record<string, string> = Object.create(null);
  for (const entry of parseEnv(contents)) {
    map[entry.key] = entry.value;
  }
  return map;
}

function isValidKey(key: string): boolean {
  return /^[A-Za-z_][A-Za-z0-9_]*$/.test(key);
}

function parseValue(raw: string, line: number): string {
  if (raw.length === 0) return "";

  const quoteChar = raw[0];
  if (quoteChar === '"' || quoteChar === "'") {
    const end = raw.lastIndexOf(quoteChar);
    if (end <= 0) {
      throw new ParseError(line, "unterminated quoted value");
    }
    const inner = raw.slice(1, end);
    // Single quotes are literal; double quotes honour escape sequences.
    return quoteChar === "'" ? inner : unescape(inner);
  }

  // Unquoted: strip a trailing inline comment (preceded by whitespace).
  const commentAt = raw.indexOf(" #");
  const value = commentAt >= 0 ? raw.slice(0, commentAt) : raw;
  return trimEnd(value);
}

function unescape(s: string): string {
  let out = "";
  for (let i = 0; i < s.length; i++) {
    const c = s[i];
    if (c === "\\") {
      const next = s[i + 1];
      switch (next) {
        case "n":
          out += "\n";
          i++;
          break;
        case "t":
          out += "\t";
          i++;
          break;
        case "r":
          out += "\r";
          i++;
          break;
        case "\\":
          out += "\\";
          i++;
          break;
        case '"':
          out += '"';
          i++;
          break;
        case undefined:
          out += "\\";
          break;
        default:
          out += "\\" + next;
          i++;
          break;
      }
    } else {
      out += c;
    }
  }
  return out;
}

/** Quote a string the way Rust's `{:?}` does, for parity in error messages. */
function quote(s: string): string {
  return JSON.stringify(s);
}

function stripCarriageReturn(s: string): string {
  return s.endsWith("\r") ? s.slice(0, -1) : s;
}

function trimStart(s: string): string {
  return s.replace(/^\s+/, "");
}

function trimEnd(s: string): string {
  return s.replace(/\s+$/, "");
}

File: rust/src/parse_env.rs
//! A small, forgiving `.env` parser.
//!
//! Supports `KEY=VALUE`, `export KEY=VALUE`, `#` comments, blank lines, and
//! single- or double-quoted values (with `\n`, `\t`, `\\`, `\"` escapes inside
//! double quotes). Line numbers are preserved for diagnostics.

/// A parsed assignment from a `.env` file.
#[derive(Debug, Clone, PartialEq)]
pub struct EnvEntry {
    pub key: String,
    pub value: String,
    pub line: usize,
}

/// An error encountered while parsing a `.env` file.
#[derive(Debug, Clone, PartialEq)]
pub struct ParseError {
    pub line: usize,
    pub message: String,
}

impl std::fmt::Display for ParseError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "line {}: {}", self.line, self.message)
    }
}

impl std::error::Error for ParseError {}

/// Parse the contents of a `.env` file.
pub fn parse(contents: &str) -> Result<Vec<EnvEntry>, ParseError> {
    let mut entries = Vec::new();
    for (idx, raw_line) in contents.lines().enumerate() {
        let line = idx + 1;
        let trimmed = raw_line.trim_start();
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }
        let trimmed = trimmed.strip_prefix("export ").unwrap_or(trimmed);
        let Some((key, rest)) = trimmed.split_once('=') else {
            return Err(ParseError {
                line,
                message: format!("missing '=' in assignment: {raw_line:?}"),
            });
        };
        let key = key.trim().to_string();
        if key.is_empty() || !is_valid_key(&key) {
            return Err(ParseError {
                line,
                message: format!("invalid variable name: {key:?}"),
            });
        }
        let value = parse_value(rest.trim(), line)?;
        entries.push(EnvEntry { key, value, line });
    }
    Ok(entries)
}

fn is_valid_key(key: &str) -> bool {
    let mut chars = key.chars();
    matches!(chars.next(), Some(c) if c.is_ascii_alphabetic() || c == '_')
        && key.chars().all(|c| c.is_ascii_alphanumeric() || c == '_')
}

fn parse_value(raw: &str, line: usize) -> Result<String, ParseError> {
    if raw.is_empty() {
        return Ok(String::new());
    }
    let bytes = raw.as_bytes();
    let quote = bytes[0];
    if quote == b'"' || quote == b'\'' {
        let closing = raw[1..].rfind(quote as char).map(|i| i + 1);
        let Some(end) = closing else {
            return Err(ParseError {
                line,
                message: "unterminated quoted value".to_string(),
            });
        };
        let inner = &raw[1..end];
        if quote == b'\'' {
            // Single quotes are literal.
            return Ok(inner.to_string());
        }
        return Ok(unescape(inner));
    }
    // Unquoted: strip a trailing inline comment (preceded by whitespace).
    let value = match raw.find(" #") {
        Some(i) => raw[..i].trim_end(),
        None => raw,
    };
    Ok(value.to_string())
}

fn unescape(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    let mut chars = s.chars();
    while let Some(c) = chars.next() {
        if c == '\\' {
            match chars.next() {
                Some('n') => out.push('\n'),
                Some('t') => out.push('\t'),
                Some('r') => out.push('\r'),
                Some('\\') => out.push('\\'),
                Some('"') => out.push('"'),
                Some(other) => {
                    out.push('\\');
                    out.push(other);
                }
                None => out.push('\\'),
            }
        } else {
            out.push(c);
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_basic_assignments() {
        let entries = parse("FOO=bar\nexport BAZ=qux\n# comment\n\nN=1").unwrap();
        assert_eq!(entries.len(), 3);
        assert_eq!(
            entries[0],
            EnvEntry {
                key: "FOO".into(),
                value: "bar".into(),
                line: 1
            }
        );
        assert_eq!(entries[1].key, "BAZ");
        assert_eq!(entries[2].line, 5); // comment + blank lines skipped
    }

    #[test]
    fn handles_quotes_and_comments() {
        let entries = parse(
            r#"A="hello world"
B='raw $VALUE'
C=plain # trailing
D="line\nbreak""#,
        )
        .unwrap();
        assert_eq!(entries[0].value, "hello world");
        assert_eq!(entries[1].value, "raw $VALUE");
        assert_eq!(entries[2].value, "plain");
        assert_eq!(entries[3].value, "line\nbreak");
    }

    #[test]
    fn rejects_malformed() {
        assert!(parse("NOEQUALS").is_err());
        assert!(parse("1BAD=x").is_err());
        assert!(parse("A=\"unterminated").is_err());
    }
}
